### DISCUSSION TITLE
Outputting a material property at the nodes of an element

### DISCUSSION THREAD

**Discussion by Aayush-T** [idaholab/moose#21594] [Q&A General] [answered]
Hi everyone,
I am trying to output a material property such as resolved shear stress (RSS) (in tensor mechanics module) at each node of an element. I tried various approaches such as: 
1. Defining RSS as an elemental variable (constant Monomial) and using Pointvalue in the postprocessor.
2. Defining RSS as a nodal variable (first order Lagrange) and using "selected_qp" in the auxkernel (throws an error).
3. Defining RSS as an elemental variable (constant Monomial) and using "selected_qp" in the auxkernel. 
4. Outputting .vtk

I am getting exactly the same values of RSS at all nodes in an element. This seems incorrect to me because the displacement differs at each node due to which shear stress should also be different. I suspect that the output is actually the Element average value in both exodus (Paraview) and csv files. I would really appreciate if someone can confirm whether the RSS should be the same at all nodes or different.
Thanks and regards,
Aayush Trivedi

**Comment by sapitts** [comment, marked as answer]
Hi @Aayush-T, 

The resolved shear stress is a material property and is thus computed at each of the quadrature points, see https://github.com/idaholab/moose/blob/4cb06f13e57bc2cccdf5f7b9439f4feb4521df66/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C#L416

Getting the different quadrature point values as output is a bit trickier. When I want to create a "pretty picture" of the resolved shear stress, I use a first order Monomial value, which paraview will interpret as a variable value, e.g. the auxvariable setting produces the following picture (from one of our regression tests)

```
  [twin_tau_4]
    order = FIRST
    family = MONOMIAL
  []
```
![example_firstMonomial_tau](https://user-images.githubusercontent.com/9342949/180102267-1cab8409-26ac-41d6-9ec2-6e52c4c834b5.png)

I am less sure of how to get the different quadrature point values in the CSV file. We do have the  [ElementVariableValue](https://mooseframework.inl.gov/source/postprocessors/ElementalVariableValue.html) postprocessor, but that will still give the average of the material property across the single element. Your `selected_qp` approach may work here

@GiudGiud do you know of a way to output the value of a material property at a specific `_qp` to a csv?

**Reply by GiudGiud** [reply]
Hello

Actually you're in luck, because the MaterialVPP is the ONLY object I know that outputs CSVs on qps
https://mooseframework.inl.gov/source/vectorpostprocessors/MaterialVectorPostprocessor.html

Guillaume

**Reply by Aayush-T** [reply]
Thank you!
